### PR TITLE
feat: add palindrome partitioning algorithm

### PR DESCRIPTION
**Pull request by sujaykaushik008**
#### Description of Change
fixed: #1009

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1319"></a>

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Well written and documented code, good work. 👍 😄
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/sujaykaushik008/C-Plus-Plus/actions

**Comment by sujaykaushik008** [comment]
> Well written and documented code, good work. 👍 😄
> Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/sujaykaushik008/C-Plus-Plus/actions

@Panquesito7 , thank you, I have made the changes requested, please confirm that they were made properly.

File: dynamic_programming/palindrome_partitioning.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    std::cout << "All tests passed successfully!\n";
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();
```
Comment (Panquesito7) [inline dynamic_programming/palindrome_partitioning.cpp:122]:
```suggestion
    test();  // execute the tests
```

File: dynamic_programming/palindrome_partitioning.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                }
+            }
+
+            return cuts[0][n - 1];
+        }
+    } // namespace palindrome_partitioning
+}  // namespace dynamic_programming
+
+/**
+ * Test Function
```
Comment (Panquesito7) [inline dynamic_programming/palindrome_partitioning.cpp:90]:
```suggestion
 * @brief Test Function
```

File: dynamic_programming/palindrome_partitioning.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+#include <iostream>   // for io operations
+#include <vector>     // for std::vector
+#include <algorithm>  // for std::min
+#include <cassert>    // for std::assert
+#include <climits>    // for INT_MAX
+
+/**
+ * @namespace dynamic_programming
+ * @brief dynamic programming algorithms
```
Comment (Panquesito7) [inline dynamic_programming/palindrome_partitioning.cpp:28]:
```suggestion
 * @brief Dynamic Programming algorithms
```

File: dynamic_programming/palindrome_partitioning.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
+/**
+ * @file
+ * @brief palindrome partitioning algorithm gives you the minimum number of partitions you need to make
```
Comment (Panquesito7) [inline dynamic_programming/palindrome_partitioning.cpp:3]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: dynamic_programming/palindrome_partitioning.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * @namespace dynamic_programming
+ * @brief dynamic programming algorithms
+ */
+namespace dynamic_programming {
+
+    /**
+     * @namespace palindrome_partitioning
+     * @brief palindrome partitioning algorithms
```
Comment (Panquesito7) [inline dynamic_programming/palindrome_partitioning.cpp:34]:
```suggestion
     * @brief Functions for [Palindrome Partitioning](https://leetcode.com/problems/palindrome-partitioning-ii/) algorithm
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: dynamic_programming/palindrome_partitioning.cpp (hunk 4 above)
Comment (sujaykaushik008) [reply dynamic_programming/palindrome_partitioning.cpp:3]: I have now added a link to geeksforgeeks site that provide proper in-depth algorithm explanation

File: dynamic_programming/palindrome_partitioning.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+
+    /**
+     * @namespace palindrome_partitioning
+     * @brief Functions for [Palindrome Partitioning] (https://leetcode.com/problems/palindrome-partitioning-ii/) algorithm
```
Comment (Panquesito7) [inline dynamic_programming/palindrome_partitioning.cpp:35]:
```suggestion
     * @brief Functions for [Palindrome Partitioning](https://leetcode.com/problems/palindrome-partitioning-ii/) algorithm
```

File: dynamic_programming/palindrome_partitioning.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+ * [Palindrome Partitioning] (https://www.geeksforgeeks.org/palindrome-partitioning-dp-17/)
+ * overall time complexity O(n^2)
+ * For example: example 1:-
+ * String : "nitik"
+ * Output : 2 => "n | iti | k"
+ * For example: example 2:-
+ * String : "ababbbabbababa"
+ * Output : 3 => "aba | b | bbabb | ababa"
+ * @author [Sujay Kaushik] (https://github.com/sujaykaushik008)
+ * @see related problem at [Leetcode] (https://leetcode.com/problems/palindrome-partitioning-ii/)
```
Comment (Panquesito7) [inline dynamic_programming/palindrome_partitioning.cpp:18]:
```suggestion
```

File: dynamic_programming/palindrome_partitioning.cpp [hunk 8]
```diff
@@ -0,0 +1,125 @@
+/**
+ * @file
+ * @brief palindrome partitioning algorithm gives you the minimum number of partitions you need to make
```
Comment (Panquesito7) [inline dynamic_programming/palindrome_partitioning.cpp:3]:
```suggestion
 * @brief Implements [Palindrome Partitioning](https://leetcode.com/problems/palindrome-partitioning-ii/) algorithm, giving you the minimum number of partitions you need to make
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: dynamic_programming/palindrome_partitioning.cpp (hunk 6 above)
Comment (sujaykaushik008) [reply dynamic_programming/palindrome_partitioning.cpp]: suggested change has been made

File: dynamic_programming/palindrome_partitioning.cpp (hunk 7 above)
Comment (sujaykaushik008) [reply dynamic_programming/palindrome_partitioning.cpp]: suggested changes have been made

**Review by Panquesito7 — APPROVED** [review]
LGTM; code looks neat, good work. 😄 👍

**Comment by sujaykaushik008** [comment]
@kvedala plse review the PR.

**Review by ayaankhan98 — APPROVED** [review]
@sujaykaushik008 Thanks for your contribution.
LGTM

**Pull request merged**